internal/metrics/prometheus/factory.go

```
@@ -148,8 +154,13 @@ func (f *Factory) Gauge(options metrics.Options) metrics.Gauge {
 		Help: help,
 	}
 	gv := f.cache.getOrMakeGaugeVec(opts, labelNames)
+	labelValues := f.tagsAsLabelValues(labelNames, tags)
+	metric, err := gv.GetMetricWithLabelValues(labelValues...)
```

IIUC WithLabelValues could panic on bad strings but GetMetricWithLabelValues would return an error. But doesn't prom library have some utility to actually sanitize the strings to something valid? If think it's better to record a slightly different label with special characters encoded than to silently ignore the data.

---

I've done some research on this topic, and here are my findings:

- I couldn't find in the Prometheus library functionality for handling string sanitization. In general this responsibility is left to the user of the library.

> // While populating dto.Metric, it is the responsibility of the
> // implementation to ensure validity of the Metric protobuf (like valid
> // UTF-8 strings or syntactically valid metric and label names). It is
> // recommended to sort labels lexicographically. Callers of Write should
> // still make sure of sorting if they depend on it.

- There are 2 options for replacing the [WithLabelValues(lvs ...string)](https://pkg.go.dev/github.com/prometheus/client_golang@v1.22.0/prometheus#CounterVec.WithLabelValues):
    - [GetMetricWithLabelValues(lvs ...string)](https://pkg.go.dev/github.com/prometheus/client_golang@v1.22.0/prometheus#CounterVec.GetMetricWithLabelValues)
        - Pros:
            - It accepts the same arguments as WithLabelValues.
        - Cons:
            - Prone to mistakes caused by an incorrect order of arguments.
    - [GetMetricWith(labels Labels)](https://pkg.go.dev/github.com/prometheus/client_golang@v1.22.0/prometheus#CounterVec.GetMetricWith)
        - Pros:
            - A more robust approach that helps prevent potential side effects and unexpected behavior.
        - Cons:
            - Requires a bit more effort to implement, because it accepts slightly different argument. 
            - Introduces performance overhead if there are many labels. 
    
    >    Note that for more than one label value, this method is prone to mistakes caused by an incorrect order of 
    >     arguments. Consider GetMetricWith(Labels) as an alternative to avoid that type of mistake. For higher label 
    >     numbers, the latter has a much more readable (albeit more verbose) syntax, but it comes with a performance  
    >     overhead (for creating and processing the Labels map).  

    IMO it is worth the effort to use **GetMetricWith**, because it is important to avoid potential unexpected side effects at all cost.

- For the UTF-8 sanitization topic, I founded 2 approaches that might solve the problem:
    - [strings.ToValidUTF8](https://pkg.go.dev/strings#ToValidUTF8): We can use it to directly replace every character that is invalid with empty space.
        ```go
        import "strings"

        sanitized := strings.ToValidUTF8(input, "")
        ```
    - [runes.ReplaceIllFormed](https://pkg.go.dev/golang.org/x/text@v0.25.0/runes#ReplaceIllFormed): We can use it for more advanced replacement of invalid characters, so that users are aware there were issues with them(we already use `golang.org/x/text` library, so we don't introduce extra dependency).
        ```go
        import (
            "golang.org/x/text/transform"
            "golang.org/x/text/runes"
            "unicode/utf8"
            "strings"
            "io/ioutil"
       )

      func sanitizeUTF8(input string) (string, error) {
            t := transform.Chain(runes.ReplaceIllFormed(), transform.Nop)
            result, _, err := transform.String(t, input)
            return result, err
        }
        ```
    IMO it is better approach to use **ReplaceIllFormed**, because:
    - We avoid silently removing characters that users might expect to see.
    - If all characters in the given label are invalid UTF-8, simply replacing them with empty strings could lead to errors. In such cases, a more advanced replacement mechanism is helpful to preserve the structure and inform users about the issue.

---

stepping through prom library I found that the only validation it seems to be doing is
```
	for _, val := range vals {
		if !utf8.ValidString(val) {
			return fmt.Errorf("label value %q is not valid UTF-8", val)
		}
	}
```
So we should indeed implement sanitizeUTF8 function. I would argue it should be the trace sanitizer because the only strings we use for metrics are service names and span names, and those better be valid UTF8 strings since we display then in the UI.

To make the sanitizer not hide/swallow the invalid runes we can replace them with hex equivalent, such that the string from your test "foo\x85bar" would render something like "foo_x85_bar".

We could also apply the same sanitizer in the prom factory, just in case, but it would be no-op if the span was already sanitized.